teps/0098-workflows.md

```diff
@@ -256,11 +302,163 @@ Why should this TEP _not_ be implemented?
+Use tekton standards for code, conduct, and contributions. Choose and document a Tekton stability level for the existing API.
+1. Update the project to make more extensive use of remote resolution. For example, we may want to keep the ability to refer to Tekton CRDs in the same
+repo using local paths, but encourage the use of the built-in hub resolvers rather than using the Pipelines as Code
+[hub resolution](https://pipelinesascode.com/docs/guide/resolver/#tekton-hubhttpshubtektondev) and the built-in git resolver rather than using
+the Pipelines as Code [remote URL syntax](https://pipelinesascode.com/docs/guide/resolver/#remote-http-url).
+1. Create a Workflow CRD that allows users to configure much of the behavior that's currently configured through annotations, such as which events
+trigger a PipelineRun and concurrency controls to apply. We may choose to rework the existing implementation so that configuring a PipelineRun
+with annotations would be implemented by creating a Workflow under the hood. We'd support the existing API (PipelineRun with annotations) for at
+least 6 months, and then we can reevaluate how to reduce duplication and create a more consistent API while still maintaining an easy setup experience
+for users who want to configure workflows only via their Git repo rather than on the Kubernetes cluster.
+1. Set a long term goal of updating the project to use Triggers. The project used to be based on Triggers (for more info, see the 
+[latest branch](https://github.com/openshift-pipelines/pipelines-as-code/tree/release-0.5.0/config) using a Triggers backend),
+but migrated to a custom implementation due to several of the [pain points](#use-of-triggers) identified above.
+We'll have to fix these pain points before reworking the implementation to use Triggers, and the refactoring may be challenging to do incrementally
+or not worth the time required to do so. We'll have to weigh the difficulty of refactoring against the difficulty of maintaining
+similar code in both projects.
+1. Explore building extensibility mechanisms into the project. It's not yet clear what this would look like.
+
+### Adopt Pipelines as Code alongside Workflows
+
+In this solution, the Tekton community would adopt and maintain the Pipelines as Code project, but not
+use it as a Workflows replacement. Instead, it could serve as an example of how to build a CI/CD platform
+on top of Tekton Pipelines. Pipelines as Code can continue to focus on an approach where Pipeline definitions
+are stored in the repository that is the subject of a CI/CD workflow, and can be tested by creating a pull request
+on the repository, while Workflows can focus on an approach more similar to other Tekton projects,
+where a cluster operator installs the Workflows project and configures repo connections and an app developer
+defines Workflows by creating CRDs on a cluster.
```

Comment:
This distinction feels crucial but I'm not sure I quite get it from a UX perspective...In the Workflows scenario, how generally might app developers create CRDs on the cluster?

Reply:
Right now in PAC the user stores a Pipeline definition in a specific folder in the repo, which the PAC controller automatically applies to the cluster. The approach we take with Pipelines and Triggers (which "workflows" would be similar to) is that a dev applies a yaml definition to their cluster via kubectl (and might use more tools like kustomize). Maybe the ops team has some scripts that keep the infrastructure repo/folder in sync with the cluster, or they could use a tool like fluxcd to do so.

Reply:
It goes a bit more than being stored in the same repository. It means as well that the definition of the pipelinerun is at the same time of when the event is created.

When I say event i mean a pull or a push request here, if a allowed user send a pull request modifying the PipelineRun in .tekton directory and another user send another pull request from the .tekton directory they are both going to have their change tested independently with no conflicts.

Reply:
> When I say event i mean a pull or a push request here, if a allowed user send a pull request modifying the PipelineRun in .tekton directory and another user send another pull request from the .tekton directory they are both going to have their change tested independently with no conflicts.

I think this use case is possible with pipelines + remote resolution today as well?

Reply:
@dibyom I think this is possible but significantly more challenging using only existing tekton features. My understanding of the feature being described is that a trusted user can open a PR with a pipelinerun, and that pipelinerun will be run as part of CI on that pull request even before it is merged. To achieve this with pipelines as it currently exists, you'd have to set up a webhook that receives events related to open PRs, filter for PRs that change files in the .tekton directory and were opened by trusted users (possibly having to write a custom interceptor?), and have your CI pipeline run the contents of the PR. Doable but definitely a lot more work than using PAC.

Reply:
Added a "user experience" section that addresses some of this discussion, PTAL

Reply:
@lbernick The point I was trying to make is that it is possible to have the workflows yaml not be in the git repo (e.g. a CRD that is applied to the cluster) and still support the use case of testing out changes to the pipeline as long as the pipeline definition is fetched from git via remote resolution

(A roughly example of what this could look like is https://github.com/dibyom/workflows/blob/main/minimal.yaml)

Reply:
yes, the minimal.yaml @dibyom pointed would work as long the workflow crd is pre-deployed on cluster. 

But I think having that CRD directly in the git repo close where the code is makes it easier to manage/versioning/lifecycle. 

Some users may want to disassociate the pipeline from the repository configuration but some users (most of our users actually) would have the whole definition directly int he git repository.

This goes with the same promise as keeping it simple to manage like travis/circle/github does it.